### ISSUE TITLE
[data grid] Unexpected keyboard tab navigation behaviour within cells

### ISSUE THREAD

**Issue by sd-danhill** [mui/mui-x#14205]
### Steps to reproduce

Link to live example: https://codesandbox.io/s/mui-tab-nav-issue-fsj87n?file=/src/Demo.tsx

Steps:
1. Select a cell in column `Quantity`
2. Use `Tab` to cycle through cells
3. `Tab` through until you reach end of row

https://github.com/user-attachments/assets/073f7b70-31fe-4467-a4a9-012a57753a2c

### Current behavior

When using tab on the last cell, the cell that is selected next looks to be the first cell in the virtualised window

### Expected behavior

When reaching the last cell in row, expected behaviour is the next cell that should be selected when tabbing should be first column on new row

### Context

Standard tab navigation

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Search keywords**: keyboard tab navigation
**Order ID**: 78997

**Comment by michelengelen** [comment]
That's correct @sd-danhill ... but mainly because we do not support navigating the table with tab. This is only possible since you applied `slotProps.cell.tabIndex: 0`

This opens another question though: We do not have a way to move to the first cell in the next row when reaching the last cell.

We could probably introduce a new prop to allow for such a thing and capture it in the navigation.

here is a diff with a working solution:
```diff
diff --git a/packages/x-data-grid/src/hooks/features/keyboardNavigation/useGridKeyboardNavigation.ts b/packages/x-data-grid/src/hooks/features/keyboardNavigation/useGridKeyboardNavigation.ts
index 1854e8a13..daa41ec01 100644
--- a/packages/x-data-grid/src/hooks/features/keyboardNavigation/useGridKeyboardNavigation.ts
+++ b/packages/x-data-grid/src/hooks/features/keyboardNavigation/useGridKeyboardNavigation.ts
@@ -39,21 +39,28 @@ const getLeftColumnIndex = ({
   firstColIndex,
   lastColIndex,
   direction,
+  jumpToPreviousRow = true,
 }: {
   currentColIndex: number;
   firstColIndex: number;
   lastColIndex: number;
   direction: 'rtl' | 'ltr';
+  jumpToPreviousRow?: boolean;
 }) => {
   if (direction === 'rtl') {
     if (currentColIndex < lastColIndex) {
       return currentColIndex + 1;
     }
-  } else if (direction === 'ltr') {
+    return jumpToPreviousRow ? firstColIndex : null;
+  }
+
+  if (direction === 'ltr') {
     if (currentColIndex > firstColIndex) {
       return currentColIndex - 1;
     }
+    return jumpToPreviousRow ? lastColIndex : null;
   }
+
   return null;
 };

@@ -62,21 +69,28 @@ const getRightColumnIndex = ({
   firstColIndex,
   lastColIndex,
   direction,
+  jumpToNextRow = true,
 }: {
   currentColIndex: number;
   firstColIndex: number;
   lastColIndex: number;
   direction: 'rtl' | 'ltr';
+  jumpToNextRow?: boolean;
 }) => {
   if (direction === 'rtl') {
     if (currentColIndex > firstColIndex) {
       return currentColIndex - 1;
     }
-  } else if (direction === 'ltr') {
+    return jumpToNextRow ? lastColIndex : null;
+  }
+
+  if (direction === 'ltr') {
     if (currentColIndex < lastColIndex) {
       return currentColIndex + 1;
     }
+    return jumpToNextRow ? firstColIndex : null;
   }
+
   return null;
 };

@@ -575,10 +589,11 @@ export const useGridKeyboardNavigation = (
             lastColIndex,
             direction,
           });
+          const canJumpToNextRow = rightColIndex === 0 && rowIndexBefore < lastRowIndexInPage;
           if (rightColIndex !== null) {
             goToCell(
               rightColIndex,
-              getRowIdFromIndex(rowIndexBefore),
+              getRowIdFromIndex(rowIndexBefore + (canJumpToNextRow ? 1 : 0)),
               direction === 'rtl' ? 'left' : 'right',
             );
           }
@@ -592,10 +607,11 @@ export const useGridKeyboardNavigation = (
             lastColIndex,
             direction,
           });
+          const canJumpToPreviousRow = leftColIndex === lastColIndex && rowIndexBefore > firstRowIndexInPage;
           if (leftColIndex !== null) {
             goToCell(
               leftColIndex,
-              getRowIdFromIndex(rowIndexBefore),
+              getRowIdFromIndex(rowIndexBefore - (canJumpToPreviousRow ? 1 : 0)),
               direction === 'rtl' ? 'right' : 'left',
             );
           }
```

WDYT @mui/xgrid ?

**Comment by arminmeh** [comment]
>I think the ARIA spec degrades usability in many cases, but there's indeed no way around that if we want to be ARIA compliant.
>
>We could however add an opt-in option to enable cycling cells through tab, it's a frequently requested feature and in some cases (excel-like applications) it makes sense to enable it.

In this case, I see that it can be useful not to have tabbing through the grid
If the user is not interested in the grid content and wants to scan the rest of the page, the user will be forced to use something like heading shortcuts or similar to skip the grid if the tab key used to navigate the grid

With our current setup in place, once you are focused on the grid you use other keys to interact with the grid and
if you are not interested in the grid data you tab to go further

**Comment by sd-danhill** [comment]
Thanks all. I've upvoted the feature request.

IMO I do agree with @romgrk, I understand the desire to be ARIA compliant but when majority of the data grids use tab navigation and like us, have been migrating to MUI, when our users ask why they can't tab navigate like they can on other products they're used to using, it's not really an easy sell to tell them that our new grid is ARIA compliant and the others maybe aren't.

It's going to have to be a feature I'm going to have to "hack" into the grid.

I'd really love an opt-in way to enable this.

I've also found this ticket which seems related:

https://github.com/mui/mui-x/issues/4658

**Comment by michelengelen** [comment]
That's right @sd-danhill ... the issue is indeed related and provides some additional information!

I'll close this in favor of #9415

**Comment by arminmeh** [comment]
I agree that there might be scenarios where being 100% aligned with the ARIA specs is not important.
So, having a flag to opt-in into a different scenario is fine for me as well. We should properly document what the usage of the flag means for the accessibility score
